Added feature to print contents from file

tuggan opened the pull request:
I've added a function to print contents from a file to the statusbar (with newlines removed).
---
I implemented this to print information from /sys/ and /proc/ in my statusbar but it can print information from any readable file.

Airblader commented:
#66, #89, #19, #313

eplanet commented:
I think the feature is both lightweight and independent enough to guarantee not breaking anything. Considered it's the fifth time it's been implemented and proposed it means people want it, wouldn't it be time to reconsider integrating it? That PR is quite clean, with unit test and so on...

stapelberg commented:
Yeah, upon reconsideration, this actually seems like a relatively clean and light-weight way to extend i3status without the complexity of process management.

stapelberg commented on src/print_file_contents.c:
```diff
@@ -0,0 +1,72 @@
+    struct stat st;
+    const bool exists = (stat(path, &st) == 0);
+
+    int chars = ((max_chars + 1) < 4096 ? max_chars + 1 : 4096);
```
Why not just use max_chars? Allocating less than a page should be fine, no?

stapelberg commented on src/print_file_contents.c:
```diff
@@ -0,0 +1,72 @@
+    const char *walk = format;
+    char *outwalk = buffer;
+
+    /* Check if file exists */
```
It would be more elegant to derive file existance from the errno value of the `open()` in line 36 (`ENOENT`).

stapelberg commented on src/print_file_contents.c:
```diff
@@ -0,0 +1,72 @@
+    const bool exists = (stat(path, &st) == 0);
+
+    int chars = ((max_chars + 1) < 4096 ? max_chars + 1 : 4096);
+    char *filebuff = scalloc(chars * sizeof(char));
```
Nit: please name this “buf” for consistency with the rest of i3status

stapelberg commented on src/print_file_contents.c:
```diff
@@ -0,0 +1,72 @@
+    int n = -1;
+    if ((fd = open(path, O_RDONLY)) > -1) {
+        n = read(fd, filebuff, chars - 1);
+        if (n != -1)
```
nit: please use curly braces here (and below). we are working towards having them everywhere.

stapelberg commented on i3status.c:
```diff
@@ -491,6 +491,17 @@ int main(int argc, char *argv[]) {
         CFG_END()};
 
+    cfg_opt_t read_opts[] = {
+        CFG_STR("format", "%content", CFGF_NONE),
```
Can you please introduce a format for the error message, too? That way, users can customize it.

tuggan commented:
I'm willing to make the changes needed

tuggan commented:
I've added the changes mentioned, the build is failing because not all files are formatted correctly.

When i ran a format on my directory three files came out changed so instead of commiting just a formatting change i thought I'd hear how you want to resolve this?
	modified:   src/print_file_contents.c
	modified:   src/print_load.c
	modified:   src/print_wireless_info.c

stapelberg commented:
> When i ran a format on my directory three files came out changed so instead of commiting just a formatting change i thought I'd hear how you want to resolve this?

Not entirely sure I understand your question. In the i3 projects, we use automated code formatting before committing changes.

Speaking of which, could you squash your commits please? We like to have a clean git history. Thanks!

tuggan commented:
> 
> 
> > When i ran a format on my directory three files came out changed so instead of commiting just a formatting change i thought I'd hear how you want to resolve this?
> 
> Not entirely sure I understand your question. In the i3 projects, we use automated code formatting before committing changes.

The question is no longer relevant :)

 
> Speaking of which, could you squash your commits please? We like to have a clean git history. Thanks!

I hope this is what you're looking for

stapelberg commented:
Yeah, thanks, the commit looks good now.

However, the build still fails (see the travis run). Can you add the following hunk to your commit please?

```diff
diff --git i/Makefile.am w/Makefile.am
index 7d8b5a9..bb251f0 100644
--- i/Makefile.am
+++ w/Makefile.am
@@ -67,6 +67,7 @@ i3status_SOURCES = \
 	src/print_time.c \
 	src/print_volume.c \
 	src/print_wireless_info.c \
+	src/print_file_contents.c \
 	src/process_runs.c \
 	src/pulse.c
 
```

tuggan commented:
Added, and build now succeeds.

stapelberg commented:
Thanks!
